iroh/src/protocol.rs

```diff
@@ -350,8 +369,64 @@ async fn handle_connection(incoming: crate::endpoint::Incoming, protocols: Arc<P
         warn!("Ignoring connection: unsupported ALPN protocol");
         return;
     };
-    if let Err(err) = handler.accept(connecting).await {
-        warn!("Handling incoming connection ended with error: {err}");
+    match handler.on_connecting(connecting).await {
+        Ok(connection) => {
+            if let Err(err) = handler.accept(connection).await {
+                warn!("Handling incoming connection ended with error: {err}");
+            }
+        }
+        Err(err) => {
+            warn!("Handling incoming connecting ended with error: {err}");
+        }
+    }
+}
+
+/// Wraps an existing protocol, limiting its access,
+/// based on the provided function.
+#[derive(derive_more::Debug, Clone)]
+pub struct AccessLimit<P: ProtocolHandler + Clone> {
+    proto: P,
+    #[debug("limiter")]
+    limiter: Arc<dyn Fn(NodeId) -> bool + Send + Sync + 'static>,
```

Comment:
Not sure if this was discussed before: will we get folks upset that they can't do async stuff here?  I think this itself runs in an async context.

Reply:
Of course we/they can just make an `AsyncAcessLimit` which is probably nicer than forcing everyone to use async.

Reply:
Sure, why not.  Don't think there'd be an objection to this.